File: src/events/release.rs
use crate::errors::Error;
use crate::events::Event;
use crate::events::base::{Release, User, WebhookMessage};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct ReleaseEvent {
    pub action: String,
    pub release: Release,
    pub sender: User,
}

impl Event for ReleaseEvent {
    fn handle(&self) -> Result<Option<WebhookMessage>, Error> {
        if self.action != "published" {
            return Ok(None);
        }

        Ok(Some(WebhookMessage {
            content: format!(
                "[{}](<{}>) published release [{}](<{}>)",
                self.sender.login,
                self.sender.html_url,
                self.release.tag_name,
                self.release.html_url
            ),
            username: self.sender.login.clone(),
            avatar_url: self.sender.avatar_url.clone(),
        }))
    }
}

File: parser/release.go
package parser

import (
	"fmt"
	"webhook/structs"
)

type release struct {
	Action  string          `json:"action"`
	Release structs.Release `json:"release"`
	Sender  structs.User    `json:"sender"`
}

func (e *release) handle() (*structs.Webhook, error) {
	if e.Action != "published" {
		return nil, nil
	}

	return &structs.Webhook{
		Content: fmt.Sprintf(
			"[%s](<%s>) published release [%s](<%s>)",
			e.Sender.Login,
			e.Sender.HtmlUrl,
			e.Release.TagName,
			e.Release.HtmlUrl,
		),
		Username:  e.Sender.Login,
		AvatarUrl: e.Sender.AvatarURL,
	}, nil
}
